elixir/apps/domain/lib/domain/gateways/group.ex

```diff
@@ -4,6 +4,7 @@ defmodule Domain.Gateways.Group do
   schema "gateway_groups" do
     # TODO: name
     field :name_prefix, :string
+    field :routing, Ecto.Enum, values: ~w[managed self_hosted stun_only turn_only]a
```

Comment:
I think before naming enum values we should think about use cases for it:
1. Any
2. Stun Only
3. Self-Hosted Relays only
4. Stun on Self-Hosted Relays only

I'm not sure why somebody will want a TURN only when STUN fails back to TURN if it fails. This option should also be some sort of enterprise feature because we will relay a ton of traffic. cc @jamilbk @conectado 

```suggestion
    field :routing, Ecto.Enum, values: ~w[all self_hosted_only stun_only turn_only]a
```

Reply:
The enum values I listed here are the ones that are explicitly called out in the in issue: https://github.com/firezone/firezone/issues/2472

Reply:
TURN always fails back to STUN not the other way around and that's why normally one would only use TURN URLs.

One would want STUN only as a "hole-punched only" connection.

Then we would have relayed-only connections which do require more configuration than using a different URL same for relayed only on self-hosted.

I do think we shouldn't expose the STUN/TURN details to the user, those are implementation details and we might want to change the standard in the future, the names I propse are:
* Any
* Hole-punched only[ex-STUN only]
* Relayed only
* Relayed only on self-hosted

Reply:
The STUN/TURN details are not exposed to the users (for the most part), they were only being used for the enum values.  The values currently being exposed to the user are as follows:

* Managed Relays
* No Relays (STUN-only)

We can drop the `STUN-only` part of the last one if we want. 

@conectado, perhaps I'm missing some details around STUN/TURN, but if I'm thinking of this routing option from the perspective of "which relays do I want my data potentially flowing through?" then wouldn't it come down to 3 options:

* Managed Relays
* Self-hosted Relays
* No Relays

You might be able to say an option of 'Any' could also make sense, but I'm having a difficult time thinking of a situation where where an account would have self-hosted relays and not want to use them exclusively.  It feels like if a customer takes the time to self-host a relay they'd most likely only want to be using the relays they setup, otherwise, what advantage would it be for the customer to set the relays up in the first place?  This also gave me another thought: If a customer has self-hosted relays, should we allow them to use the managed relays?

Finally, I'm curious if there is a practical case for needing `Relayed-only`?  I'm having a hard time thinking of when a customer would want to force all of their traffic through Relays intentionally.  But again, perhaps there are some details I'm missing around STUN/TURN.

Reply:
@bmanifold I feel like "No Relays (STUN-only)" is a bit ambiguous naming. From it you might assume only a direct connection will be used (eg. you opened a port and know the gateway IP), while IRL relays will be used for hole punching, just not for relaying the traffic.  

> You might be able to say an option of 'Any' could also make sense, but I'm having a difficult time thinking of a situation where where an account would have self-hosted relays and not want to use them exclusively.

Any large enough account that has strict requirements to access one of the environments but has other less relaxed environments. For example, medical professionals accessing on-site MRI's image storage using only self-hosted relays or no relays at all (IRL they want not just STUN but only direct connections; one such customer is in our Slack already) and then having something like other cloud environment where they don't want to pay for the traffic bill or maintain relays there. 

> Finally, I'm curious if there is a practical case for needing Relayed-only? I'm having a hard time thinking of when a customer would want to force all of their traffic through Relays intentionally. But again, perhaps there are some details I'm missing around STUN/TURN.

+1. I don't think this option makes a ton of sense, it will double the traffic for self-hosted relays or for us, while not giving any benefits to the customers.

Reply:
WDYT about:

1. Use Firezone- or self-hosted Relays (STUN and TURN)
2. Use Firezone- or self-hosted Relays only for hole-punching (STUN-only)
3. Only use self-hosted Relays (STUN and TURN)


The reason why I extended (1) and (2) is to cover a long-term case like this one:

A client with offices in multiple countries but no Firezone-hosted relays nearby one of the offices (eg. Sough Africa). They can manage their own relay set nearby to that site and Firezone will automatically use it if it's the closest one to the users, while the rest of the world will use Firezone-hosted ones. 

This will be also easier to implement because this is how we pick relays already.

Reply:
If we want to have more content to explain the options we can also reuse similar components here:
<img width="700" alt="Screenshot 2023-11-09 at 10 33 23" src="https://github.com/firezone/firezone/assets/1877644/854911e8-c88d-455f-b910-66dd6acb09b1">

Reply:
We aren't launching with self-hosted Relays, so it doesn't make sense to list them in the menu yet. The only options would then be `Use Firezone-managed Relay Network` and `Peer to peer connections only`.

I think this could be simplified to just a radio button too:

![](https://share.balsamiq.com/c/sG3gaW5CxqLDTLhpo8xRFb.png)

Reply:
> I feel like "No Relays (STUN-only)" is a bit ambiguous naming. From it you might assume only a direct connection will be used (eg. you opened a port and know the gateway IP), while IRL relays will be used for hole punching, just not for relaying the traffic.

If someone is thinking about the low level details then I would agree this could be slightly ambiguous, however, if you're looking at it from the simpler perspective of "is my data flowing through a relay?", then I don't think it's really ambiguous at all.  I feel like we should try and optimize for the more simplistic view to take some of the cognitive load off of the user.

>>You might be able to say an option of 'Any' could also make sense, but I'm having a difficult time thinking of a situation where where an account would have self-hosted relays and not want to use them exclusively.

> Any large enough account that has strict requirements to access one of the environments but has other less relaxed environments. For example, medical professionals accessing on-site MRI's image storage using only self-hosted relays or no relays at all (IRL they want not just STUN but only direct connections; one such customer is in our Slack already) and then having something like other cloud environment where they don't want to pay for the traffic bill or maintain relays there.

I think one of the things that was implied in my statement that I probably should have made explicit is "I'm having a difficult time thinking of a situation where where an account would have self-hosted relays and not want to use them exclusively (for a given site)".  I totally agree there could be situations where a customer might want to use a different relay strategy for different sites, but in a given site, I'm having a difficult time thinking of why they would need an 'Any' option.

I think the easiest thing to do right now is not worry too much about it as we've only got two options we're going to present to the user anyway, so I propose narrowing the Enum list down to `managed` and `stun_only`.  Then change the UI to use the radio buttons as shown in the wireframe above.

Does that work for everyone?

Reply:
@bmanifold Yeah, makes sense. Let's move forward with that for launch and use feedback to inform this debate further.